include/circt/Dialect/SystemC/SystemCTypes.h

```diff
@@ -32,6 +48,244 @@ struct PortInfo {
 /// Get the type wrapped by a signal or port (in, inout, out) type.
 Type getSignalBaseType(Type type);
 
+//===----------------------------------------------------------------------===//
+// Integer types
+//===----------------------------------------------------------------------===//
+
+/// This provides a common base class for all SystemC integers.
+/// It represents the sc_value_base class described in IEEE 1666-2011 §7.4.
+/// Note that this is an abstract type that cannot be instantiated but can only
+/// be used as a base class and for type checks.
+class ValueBaseType : public Type {
+public:
+  static bool classof(Type type) {
+    return type.isa<SignedType>() || type.isa<UnsignedType>() ||
+           type.isa<IntBaseType>() || type.isa<UIntBaseType>() ||
+           type.isa<BigIntType>() || type.isa<BigUIntType>() ||
+           type.isa<IntType>() || type.isa<UIntType>();
+  }
+
+protected:
+  using Type::Type;
+};
```

Comment:
I haven't looked at this type of thing in a while. However, is there some way that more of this can be specified inside ODS as opposed to rolling custom classes? I see that the `*BaseType`s are all in ODS as are type constraints that relate to these, but are not these classes.

Reply:
ODS does not support subtyping between custom types (all types in ODS have to inherit from `mlir::Type` directly). As an alternative we can define constraints in ODS like `AnyTypeOf<[SignedType, UnsignedType, IntBaseType, ...]>` and use them in ODS but not in C++ code. If we want to check in C++ if we have a SystemC value type we can just do `type.isa<ValueBaseType>()` in the current implementation, when switching to pure ODS we have to do `type.isa<SignedType, UnsignedType, IntBaseType, UIntBaseType, BitIntType, BitUIntType, IntType, UIntType>()` or define a helper function in `SystemCTypes.h` to do that. Same applies for the other subtype relations. It also allows to define member functions in the common base class that work on all subtypes (e.g., we have a `getSignalBaseType` helper function in the header file to operate on all port and signal types; if those would have a common base type it could be a member function of that instead; seems to come down to procedural vs object-oriented coding style)